content/blog/2021-09-22-refactor.md

```diff
@@ -0,0 +1,163 @@
+nodes/leaves, which are all defined in the
+[Python's ASDL](https://docs.python.org/3/library/ast.html#abstract-grammar).
+
+## Refactoring source code through ASTs
+
+What [refactor](https://github.com/isidentical/refactor) does is that, it
+enables the possibility of simply matching ASTs, and then transforming them to
+different ones with doing as little as possible in terms of touching the other
+parts of the source code.
+
+```python
+foo = [
+    1,
+    2
+]
+
+foo_2 = ['a', *foo]
+
+if foo[0] >= 1:
+    assert secrets.get("foo") == foo
+```
+
+As a simple demo, let's try to find and replace all usages of `foo` variable
+name with `bar`, but without actually changing things inside strings or partial
+matches like `foo_2`.
+
+```python
+import ast
+import refactor
+```
+
+The first thing we need to do is define a rule. Each rule is a class that
+defines a single entrypoint (`match())`), takes AST nodes from the tree, and
+either rejects them (via raising an `AssertionError` or just returning `None`)
+or accepts them (via returning a `refactor.Action`).
+
+```python
+class ReplaceFoo(refactor.Rule):
+
+    def match(self, node):
+```
+
+What we do in the `match()` method is simple, we will look for all `Name`s
+(which is what the actual identifier is wrapped when used in the contexts
+above), and check whether the `node.id` is `foo` or not.
+
+```python
+        assert isinstance(node, ast.Name)
+        assert node.id == "foo"
+```
+
+If any of these assertions fail, the function will terminate at that point and
+the engine will move to the next node in the tree. But if we succeed, then it
+means that we have a match, and we should return some sort of an action. The
+simplest thing we can use is just returning a `refactor.ReplacementAction` which
+takes this node and replaces it with the given argument.
+
+```python
+        return refactor.ReplacementAction(
+            node,
+            ast.Name("bar", node.ctx)
+        )
+```
+
+And this is it. If we want to try it, we can simply create a CLI application out
+of our refactoring rules via `refactor.run()`.
+
+```python
+if __name__ == "__main__":
+    refactor.run(rules=[ReplaceFoo])
+```
+
+And if we run it on the file above, we will get this diff:
+
+```diff
+@@ -1,9 +1,9 @@
+
+-foo = [
++bar = [
+     1,
+     2
+ ]
+
+-foo_2 = ['a', *foo]
++foo_2 = ['a', *bar]
+
+-if foo[0] >= 1:
+-    assert secrets.get("foo") == foo
++if bar[0] >= 1:
++    assert secrets.get("foo") == bar
+```
+
+Which indicates all of the usages against `foo` is replaced, but items like
+`foo_2` or `"foo"` is still there as we have expected.
+
+## Going Deeper
+
+Obviously not all refactorings are as easy as this, so refactor is equipped with
+more features like different actions, and more advanced features like observers
+and representatives for context manager. If you are curious about them, be sure
+to check out it's [documentation!](https://refactor.readthedocs.io/en/latest/)
```

Comment:
```suggestion
Obviously not all refactorings are as simple as this, so `refactor` is equipped
with more features like different actions, observers and representatives for
context manager. If you are curious about these and more advance features, be
sure to check out the
[`refactor` documentation](https://refactor.readthedocs.io/en/latest)!
```

Reply:
will correct locally with other changes ->advanced

Reply:
fml today not my day for typos. edited.